### PR TITLE
RFC: functional state setter for selectors

### PR DESCRIPTION
**Pull request by anmonteiro**
For read-write selectors, change the setter to a function of the old value, to match what's done elsewhere in `useRecoilCallback`,  `useRecoilState` and `useSetRecoilState`.

### CODE REVIEW

File: src/Recoil__React.re [hunk 1]
```diff
@@ -16,10 +16,8 @@ external useRecoilState:
   Recoil__Value.readWrite('value) => ('value, ('value => 'value) => unit) =
   "useRecoilState";
 
-type value('a) = 'a;
```
Comment (anmonteiro) [inline src/Recoil__React.re:19]: this is just drive-by cleanup, which also removes this extraneous type from the public API.

Comment (bloodyowl) [reply src/Recoil__React.re:19]: I'm wondering if I didn't use that on purpose so that it forces BuckleScript not to curry if the value is a function, but I'm not sure anymore 😄.

Comment (anmonteiro) [reply src/Recoil__React.re:19]: Looking at the code generated by tests it doesn't seem like that's the case. See e.g. `atom4`, just generates the following JS:

```js
  var atom4$1 = Recoil.useRecoilValue(atom4);
```

Comment (bloodyowl) [reply src/Recoil__React.re:19]: Alright, looks ok!

**Comment by bloodyowl** [comment]
Thanks! Published in v0.10.0
